### ISSUE TITLE
Better documentation of `tensor_contract`

### ISSUE THREAD

**Issue by arnelg** [qutip/qutip#385]
The usage of the `tensor_contract` function is explained in [QuTiP Example: Superoperators, Pauli Basis and Channel Contraction](http://nbviewer.ipython.org/github/qutip/qutip-notebooks/blob/master/examples/example-superop-contract.ipynb).

It is explained in terms of the graphical notation in http://arxiv.org/abs/1111.6950. However, there is no figure explaining how the "tensor wires" are numbered. In the notebook it says "Numbering the tensor wires from 0 to 3, this corresponds to a tensor_contract argument of (1, 3)." But it is not clear what the convention is... does the numbering follow in an anticlockwise fashion, starting with 0 at the top left?

A figure in the notebook showing the numbering convention would be extremely helpful.

**Comment by cgranade** [comment]
I agree, the current documentation isn't clear. I'll work on getting better documentation up soon. Thanks for the report!

**Comment by cgranade** [comment]
Looking into the issue in more detail, I think that the documentation wasn't just misleading but wrong: because QuTiP uses the column-stacking notation, the order of ``dims`` for ``type='super'`` doesn't have the same meaning as for ``type='oper'`` or for vectors. In particular, the column-stacking basis maps |i><j| to |j> |i>, such that the order of the tensor indices for superoperators is swapped from the ordering ``[left, right]`` that is used by QuTiP's ``dims`` notation. I'll break this out into another Issue, but I think the simplest way of handling it is to keep ``dims`` the same as it is now (I don't want to break that, for crying out loud), but to transform the arguments to ``_tensor_contract_dense`` if the order of ``dims`` doesn't match the order that we need to pass to ``np.reshape``.

**Comment by IIAOPSW** [comment]
This problem has since gotten worse. The linked notebook is no longer available. I'm going to update the documentation with the numbering convention.

**Comment by veronikakurth** [comment]
Since this issue had been created, I believe [this documentation](https://qutip.readthedocs.io/en/qutip-5.0.x/guide/guide-super.html) has evolved and been improved quite a lot. 

To my understanding, the main point of the author of this issue was having an additional explanation to the tensor wires in terms of Penrose notation [here](https://qutip.readthedocs.io/en/qutip-5.0.x/guide/guide-super.html#reduced-channels). Currently, _Wood diagrams_ are referred to in the tutorial. As I understand, those are an extension to Penrose diagrams. 

@pmenczel @Ericgig Do you think the issue is still relevant and something concrete has to be done in order to enhance the documentation? It's a bit difficult for me to judge and the current state of the documentation looks very good to me, so I'd appreciate your opinion.

**Comment by pmenczel** [comment]
@veronikakurth [This is the notebook](https://github.com/qutip/qutip-notebooks/blob/master/examples/superop-contract.ipynb) that was linked originally. As far as I can tell, this notebook was migrated into the users guide (which you linked) without any changes. In particular, the text that OP found unclear is still exactly the same, and the numbering convention is not explained.

If there's still anyone around who understands what `tensor_contract` does and how the indices used in this function correspond to the `dims`, it would be very useful if they could update the documentation to explain that. (I do not understand it and would need to read the source code and try examples if I wanted to understand it.)

**Comment by veronikakurth** [comment]
> [@veronikakurth](https://github.com/veronikakurth) [This is the notebook](https://github.com/qutip/qutip-notebooks/blob/master/examples/superop-contract.ipynb) that was linked originally. As far as I can tell, this notebook was migrated into the users guide (which you linked) without any changes. In particular, the text that OP found unclear is still exactly the same, and the numbering convention is not explained.
> 
> If there's still anyone around who understands what `tensor_contract` does and how the indices used in this function correspond to the `dims`, it would be very useful if they could update the documentation to explain that. (I do not understand it and would need to read the source code and try examples if I wanted to understand it.)

Thank you very much for the prompt feedback! I see. Then I guess this issue should stay open. However, since it’s now related purely to the documentation, the labelling could be changed from ‘code’ to ‘docs’ for it to be found easier when focusing on improving the docs. Could you please take care of it?

**Comment by IIAOPSW** [comment]
> If there's still anyone around who understands what `tensor_contract` does and how the indices used in this function correspond to the `dims`, it would be very useful if they could update the documentation to explain that. (I do not understand it and would need to read the source code and try examples if I wanted to understand it.)

With the caveat its been a very long time since I've had anything to do with this project and may be imperfectly remembering some details, tensor contract is a generalization of matrix multiplication. 

In normal matrix multiplication there is one index/dimension which is summed over in the input objects (and the rest remain as indices/dimensions of the output object). Eg (A*B)_jk = sum_i A_ij * B_ki. This is a special case of tensor contraction.  

Tensor contraction takes two objects as input (with however many dimensions/indices it has) and merges them together into one output object by multiplying and summing with respect to whichever subset of indices are specified (so long as the dimensions match). Eg if I had some rank 4 tensors and contracted them together along the middle two indices I would end up with a rank 4 tensor defined by (A*B)_klmn = sum_ij A_kijl * B_mijn.  

I forget the exact syntax for the function call, but in general the dimensions of the output object are the same as the input objects but with the summed over dimensions removed. Eg if the original dimensions of my objects were [a1, a2, a3, a4] and [b1, b2, b3, b4] then the dimensions after tensor contracting the middle two indices should be [a1, a4, b1, b4].

So basically, its matrix multiplication along the axes you specify in a higher dimensional space.

**Comment by pmenczel** [comment]
Hi @IIAOPSW and thanks for coming back in with an explanation after all this time! I note that Qutip's `tensor_contract` only takes one `Qobj` as input.

I think your explanation is still essentially correct, and it just does something like $\sum_i A_{j i k i}$. However, the dimensions of the input `Qobj` will generally be something like `[[a1, a2, ..., an], [b1, b2, ..., bm]]`. Each of the input dimensions `ai` and each of the output dimensions `bi` will have a corresponding index. It is not made clear in the documentation how the input and output indices are "addressed".

(I assume that "0" up to "n-1" correspond to "a1" up to "an", and "n" up to "n+m-1" correspond to "b1" up to "bm". But it is not documented and not obvious from the code, and the discussions above make me worry that it might be more complicated somehow.)

**Comment by IIAOPSW** [comment]
> Hi [@IIAOPSW](https://github.com/IIAOPSW) and thanks for coming back in with an explanation after all this time! I note that Qutip's `tensor_contract` only takes one `Qobj` as input.

I've just looked a bit more closely at the notebook linked, and you may notice that inside the argument of both instances of "qt.tensor_contract" is "qt.to_super". Now my recollection of super-operators is that they are like two normal operators packaged together (one which left multiplies, the other which right multiplies). Eg if S is a super-operator then S*A = S_1 A S_1^\dag (or something like that)

I believe what is happening in the code is thus analogous to the way an inner product is conceptually equivalent to ordinary matrix multiplication of a vector with its own transpose conjugate. So continuing with this generalization, the code really is tensor contracting two Qobj's together, but it just so happens those two objects are the transpose conjugate of each other (as output by qt.to_super). 

This is all vague memories, but it should be easy enough to just test it with an arbitrary small matrix.

**Comment by matti-sarjala** [comment]
A small documentation improvement could be to add one explicit “index addressing” example before the Wood/Penrose diagram discussion.

The confusing point seems to be not tensor contraction in general, but this specific mapping:

- a `Qobj` has dimensions like `[[a1, a2, ..., an], [b1, b2, ..., bm]]`,
- `tensor_contract` receives integer indices,
- the docs should say exactly which integer refers to which input/output dimension.

A minimal docs example could show something like:

- for `dims=[[a1, a2], [b1, b2]]`,
- index `0` refers to `a1`,
- index `1` refers to `a2`,
- index `2` refers to `b1`,
- index `3` refers to `b2`,

or whatever the actual convention is after checking the implementation.

Then the existing example using `tensor_contract(..., (1, 3))` would become much easier to read, because the reader could see which two wires are being summed over.

I think a good first PR could avoid changing any code and only add:

1. one small diagram/table mapping `dims` entries to tensor-wire indices,
2. one tiny 2-by-2 example,
3. one sentence warning that superoperators use QuTiP’s column-stacking convention, so the wire ordering should be read from the documented convention rather than guessed from visual intuition alone.
